### PR TITLE
ARROW-15468: [R] [CI] A crossbow job that tests against DuckDB's dev branch

### PR DESCRIPTION
**Pull request by jonkeane**
This adds the ability to run the R package checks with a version of DuckDB installed from their github and adds a nightly test of that.

It also removes the workaround we had in place that builds duckdb from github when it's not available as a binary.

### CODE REVIEW

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-dev-duckdb

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-dev-duckdb

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-dev-duckdb

File: .env [hunk 1]
```diff
@@ -76,6 +76,7 @@ ARROW_R_DEV=TRUE
 R_IMAGE=ubuntu-gcc-release
 R_ORG=rhub
 R_PRUNE_DEPS=FALSE
+R_DUCKDB_DEV=FALSE
```
Comment (nealrichardson) [inline .env:79]: can we do this without adding to .env? Like, invoke docker (compose) with `-e R_DUCKDB_DEV=TRUE` in the job you care about? we don't need this as a parameter to docker-compose, just as an env var in the job.

File: docker-compose.yml [hunk 2]
```diff
@@ -1154,6 +1154,7 @@ services:
         gcc_version: ${GCC_VERSION}
         tz: ${TZ}
         r_prune_deps: ${R_PRUNE_DEPS}
+        r_duckdb_dev: ${R_DUCKDB_DEV:-}
```
Comment (jonkeane) [inline docker-compose.yml:1157]: I think that we could actually do this default setting with `R_PRUNE_DEPS` as well and remove it from `.env` if we wanted to.

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-dev-duckdb

**Comment by jonkeane** [comment]
The conda builds are addressed with what Uwe said ^^^
The autobrew has been addressed already: https://github.com/apache/arrow/commit/ee7897a3a44c8fd972544330106c3e7bbae6ba24
The valgrind has a ticket already, and is not made worse by this: https://issues.apache.org/jira/browse/ARROW-15656

File: dev/tasks/tasks.yml [hunk 3]
```diff
@@ -1185,13 +1185,22 @@ tasks:
...
+      r_tag: latest
+      flags: '-e TEST_OFFLINE_BUILD=true'
+
+  test-r-dev-duckdb:
+    ci: github
+    template: docker-tests/github.linux.yml
+    params:
+      env:
+        R_DUCKDB_DEV: TRUE
+      flags: '-e ARROW_SOURCE_HOME="/arrow" -e FORCE_BUNDLED_BUILD=TRUE -e LIBARROW_BUILD=TRUE'
```
Comment (nealrichardson) [inline dev/tasks/tasks.yml:1202]: Why are these flags necessary? That's the default for `ARROW_SOURCE_HOME` and `LIBARROW_BUILD` right? Or maybe the question is: why is this using a different `template`?

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-depsource-system test-r-dev-duckdb test-r-ubuntu-21.04 test-r-gcc-11

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-depsource-system test-r-dev-duckdb test-r-ubuntu-21.04 test-r-gcc-11

**Comment by jonkeane** [comment]
@github-actions crossbow submit test-r-dev-duckdb

**Comment by jonkeane** [comment]
The C++ / AMD64 Windows MinGW 64 C++ failure is a flake. Unless there is objection, I'll merge this later today.

**Pull request closed without merging**